assembly/src/library.rs

```diff
@@ -55,6 +55,22 @@ impl Library for MaslLibrary {
     }
 }
 
+impl Library for &MaslLibrary {
+    type ModuleIterator<'a> = Iter<'a, Module>  where Self: 'a;
+
+    fn root_ns(&self) -> &LibraryNamespace {
+        &self.namespace
+    }
+
+    fn version(&self) -> &Version {
+        &self.version
+    }
+
+    fn modules(&self) -> Self::ModuleIterator<'_> {
+        self.modules.iter()
+    }
+}
+
```

Comment:
I wonder if we can avoid the duplicated implemented of `Library` on `&MaslLibrary` and `MaslLibrary` by implementing it on `Borrow<MaslLibrary>` or maybe `AsRef<MaslLibrary>`.  I'm not sure though.

Reply:
I'm actually not sure myself if I did it the right way. Maybe @vlopes11 - could suggest?

Reply:
The implementation bellow works.  I'm sure @vlopes11 will be able to advise on the most idiomatic solution.

```rust
impl<B: core::borrow::Borrow<MaslLibrary>> Library for B {
    type ModuleIterator<'a> = Iter<'a, Module>  where Self: 'a;

    fn root_ns(&self) -> &LibraryNamespace {
        &self.borrow().namespace
    }

    fn version(&self) -> &Version {
        &self.borrow().version
    }

    fn modules(&self) -> Self::ModuleIterator<'_> {
        self.borrow().modules.iter()
    }
}
```

Reply:
We have a similar situation here:
https://github.com/0xPolygonMiden/miden-vm/blob/4195475d75ab2d586bdb01d1ff3ea2cd626eaf7b/processor/src/advice/mod.rs#L172-L231

This is a [common pattern](https://github.com/rust-lang/rust/blob/17675855094906ba90aca2f119be8fb7afc4d456/library/std/src/io/impls.rs#L17-L52). It would look like this:

```rust
impl<T> Library for &T
where
    T: Library,
{
    type ModuleIterator<'a> = T::ModuleIterator<'a>
    where
        Self: 'a;

    fn root_ns(&self) -> &LibraryNamespace {
        T::root_ns(self)
    }

    fn version(&self) -> &Version {
        T::version(self)
    }

    fn modules(&self) -> Self::ModuleIterator<'_> {
        T::modules(self)
    }
}
```

Since every ref also implements borrow, that would be a general and comprehensive implementation. The negative point of implementing traits for generics is we remove the ability from the user to customize his implementations for any type that fits the impl criteria (in this case, any reference of a `T` that implements `Library`). Given this is a simple enough example (i.e. the trait only takes immutable references of `self` that are just some getters), we shouldn't worry too much about this particular side-effect

Alternatively, we have [auto_impl](https://crates.io/crates/auto_impl) that will do the same for us via proc macros.

Reply:
Nice solution! Thanks for the explanation, this is useful to know.

Reply:
@vlopes11 - very cool! thank you!